<a href="https://colab.research.google.com/github/MuhammadShayan8401/flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/MuhammadShayan8401/flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

In [ ]:
# No code required for this section.

## My Lane (or Freestyle) and Why

I chose **Lane 2: Refresh / Content Opportunity Scoring**.

The goal of this project is to identify which content pages should be reviewed first for refresh, expansion, or optimisation. Content teams often manage thousands of pages, making it difficult to manually decide where to focus their efforts. By analysing historical search and engagement data, machine learning can provide decision-support that helps prioritise pages with the highest potential for improvement.

## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

In [ ]:
# No code required for this section.

## The Question

### Research Question

Can we predict which content pages should be prioritised for review using historical search and engagement data?

### Decision

The decision is whether a content or SEO team should spend time improving a specific page.

### Action

If a page receives a high score, the team can refresh, expand, or optimise it. If the score is low, they can monitor it or focus on other pages instead.

### Unit of Analysis

One content page.

### Output

A ranked score indicating how strongly a page should be considered for review.

### Cost of a Wrong Recommendation

A false positive wastes time and resources on pages that do not improve.

A false negative means an important page is ignored, resulting in lost traffic and missed opportunities.

### Why Machine Learning?

Machine learning can analyse many features simultaneously and identify patterns that are difficult to recognise manually, making content prioritisation more efficient.

## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [ ]:
import pandas as pd

# Load dataset from your GitHub repository (RAW URL)
url = "https://raw.githubusercontent.com/MuhammadShayan8401/flyrank-ml-internship/main/data/raw/content_refresh_anonymized.csv"

df = pd.read_csv(url)

# Dataset overview
print("=" * 50)
print("DATASET OVERVIEW")
print("=" * 50)
print(f"Total Rows              : {len(df):,}")
print(f"Total Columns           : {len(df.columns)}")
print(f"Unique Content Pages    : {df['content_id'].nunique():,}")
print(f"Unique Clients          : {df['client_id'].nunique():,}")

print("\n" + "=" * 50)
print("PERFORMANCE METRICS")
print("=" * 50)
print(f"Average Impressions (90d): {df['impressions_90d'].mean():.2f}")
print(f"Average Clicks (90d)     : {df['clicks_90d'].mean():.2f}")
print(f"Average Sessions (90d)   : {df['sessions_90d'].mean():.2f}")
print(f"Average CTR              : {df['ctr'].mean():.2f}%")
print(f"Average Position         : {df['avg_position'].mean():.2f}")

print("\n" + "=" * 50)
print("TREND DIRECTION")
print("=" * 50)
print(df["trend_direction"].value_counts())

print("\n" + "=" * 50)
print("MISSING VALUES")
print("=" * 50)
print(df.isnull().sum().sort_values(ascending=False).head(10))

DATASET OVERVIEW
Total Rows              : 30,000
Total Columns           : 44
Unique Content Pages    : 30,000
Unique Clients          : 32

PERFORMANCE METRICS
Average Impressions (90d): 5200.37
Average Clicks (90d)     : 16.10
Average Sessions (90d)   : 37.07
Average CTR              : 0.51%
Average Position         : 16.34

TREND DIRECTION
trend_direction
down      16262
stable     5962
up         4388
new        2236
flat       1152
Name: count, dtype: int64

MISSING VALUES
provider_used        21438
word_count            7699
char_count            7699
word_count_tier       7699
char_count_tier       7699
model_used            5733
trend_pct             3388
competition_level     2610
search_volume         2468
cpc                   2468
dtype: int64


## Quick Look at the Data (2–3 Real Numbers)

The starter dataset contains **30,000 content pages** from **32 anonymised clients**.

The average content page receives **5,200.37 impressions**, **16.10 clicks**, and **37.07 sessions** over a 90-day period, with an average **CTR of 0.51%**.

The performance labels also show meaningful variation across pages. Out of the 30,000 pages:

- **16,262** pages are labelled as **down**
- **5,962** are **stable**
- **4,388** are **up**
- **2,236** are **new**
- **1,152** are **flat**

These statistics suggest that the dataset contains enough variation in content performance to support a **Refresh / Content Opportunity Scoring** project. Rather than simply training a model, the goal is to build a decision-support system that helps prioritise which pages should be reviewed first based on observable search and engagement signals.

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

## Careful Words: What I Can and Can't Claim

This project provides **decision-support**, not guaranteed predictions. It identifies pages that appear more likely to benefit from review based on historical search and engagement data.

The results are **observational** and **directional**. They can help content teams prioritise pages for review, but they do not prove that refreshing a page will increase traffic or improve rankings.

This project cannot predict Google's algorithm or guarantee future search performance. External factors such as algorithm updates, competition, seasonality, and changes in user behaviour can all influence outcomes beyond what is captured in the dataset.

## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.